src/Pos/DB/Block/Block.hs

```diff
@@ -0,0 +1,302 @@
+    -- So we are loading all blocks which have difficulty ≥ targetDifficulty.
+    let targetDelta = fromIntegral depth - 1
+        targetDifficulty
+            | topDifficulty <= targetDelta = 0
+            | otherwise = topDifficulty - targetDelta
+    -- Then we load blocks starting with previous block of already
+    -- loaded block.  We load them until we find block with target
+    -- difficulty. And then we drop last (oldest) block.
+    let prev = top ^. prevBlockL
+    over _Wrapped (top :) <$>
+        loadDataWhile
+        getter
+        (\a -> a ^. difficultyL >= targetDifficulty && extraPredicate a)
+        prev
+
+-- | Load blunds starting from block with header hash equal to given hash
+-- and while @predicate@ is true.
+loadBlundsWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (Block ssc -> Bool) -> HeaderHash -> m (NewestFirst [] (Blund ssc))
+loadBlundsWhile predicate = loadDataWhile getBlundThrow (predicate . fst)
+
+-- | Load blunds which have depth less than given.
+loadBlundsByDepth
+    :: (SscHelpersClass ssc, MonadDB m)
+    => Word -> HeaderHash -> m (NewestFirst [] (Blund ssc))
+loadBlundsByDepth = loadDataByDepth getBlundThrow (const True)
+
+-- | Load blocks starting from block with header hash equal to given hash
+-- and while @predicate@ is true.
+loadBlocksWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (Block ssc -> Bool) -> HeaderHash -> m (NewestFirst [] (Block ssc))
+loadBlocksWhile = loadDataWhile getBlockThrow
+
+-- | Load headers starting from block with header hash equal to given hash
+-- and while @predicate@ is true.
+loadHeadersWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (BlockHeader ssc -> Bool)
+    -> HeaderHash
+    -> m (NewestFirst [] (BlockHeader ssc))
+loadHeadersWhile = loadDataWhile getHeaderThrow
+
+-- | Load headers which have depth less than given.
+loadHeadersByDepth
+    :: (SscHelpersClass ssc, MonadDB m)
+    => Word -> HeaderHash -> m (NewestFirst [] (BlockHeader ssc))
+loadHeadersByDepth = loadDataByDepth getHeaderThrow (const True)
+
+-- | Load headers which have depth less than given and match some criterion.
+loadHeadersByDepthWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (BlockHeader ssc -> Bool)
+    -> Word
+    -> HeaderHash
+    -> m (NewestFirst [] (BlockHeader ssc))
+loadHeadersByDepthWhile = loadDataByDepth getHeaderThrow
+
+----------------------------------------------------------------------------
+-- Initialization
+----------------------------------------------------------------------------
+
+prepareBlockDB
+    :: forall ssc m.
+       (SscHelpersClass ssc, MonadDB m)
+    => GenesisBlock ssc -> m ()
+prepareBlockDB g = do
+    putBi lastBlundFileKey (FixedSizeInt (0::Int))
+    putBlund (Left g) def
+
+----------------------------------------------------------------------------
+-- Keys
+----------------------------------------------------------------------------
+
+blockIndexKey :: HeaderHash -> ByteString
```

Comment:
I had to scroll around to remember what they mean. Could you add comments?

```
-- | DB key for getting block header

-- | DB key for location of given block
```

(I'm not sure I got them correctly.)

Also, why don't we have some `Key a = Key ByteString`? Seeing `blockIndexKey :: HeaderHash -> Key BlockHeader` or whatever would be nicer.